src/packages/@ncigdc/utils/auth/awg.js

```diff
@@ -0,0 +1,137 @@
+import { throttle } from 'lodash';
+import urlJoin from 'url-join';
+import moment from 'moment';
+
+import consoleDebug from '@ncigdc/utils/consoleDebug';
+import {
+  AWG,
+  AWG_TOKEN_EXPIRY, // defaults to 5 minutes
+  AWG_TOKEN_RENEWAL_INTERVAL, // defaults to 0, disabled
+  FENCE,
+} from '@ncigdc/utils/constants';
+
+const createSessionRenewalTimestamp = now => {
+  consoleDebug(`Creating session timestamp at ${now}`);
+  localStorage.setItem('awg_sessionLastRenewed', now);
+  return !!localStorage.awg_sessionLastRenewed;
+};
+
+const removeSessionRenewalTimestamp = () => {
+  localStorage.removeItem('awg_sessionLastRenewed');
+  return !localStorage.awg_sessionLastRenewed;
+};
+
+const resetSessionRenewalTimestamp = now => (
+  removeSessionRenewalTimestamp() &&
+  createSessionRenewalTimestamp(now)
+);
+
+/* this is meant to avoid duplicate queries, because having multiple Relay Environments
+ * means renewal race conditions not worth dealing with at this point. ¯\_(ツ)_/¯
+ */
+let queryCache = null;
+const handleQueryCache = value => {
+  // hold and use the latest response for 5 seconds;
+  queryCache = value;
+
+  setTimeout(() => {
+    queryCache = null;
+  }, 5000);
+
+  return queryCache;
+};
+
+const renewAWGToken = throttle(({
```

Comment:
Could you confirm I understand this function's intended interface correctly?

1. If the renewal succeeds, `renewAWGToken` resolves to true.
2. If the HTTP request returns 403, `renewAWGToken` resolves to false, so `checkAWGSession` also resolves to false, and the upstream caller then redirects to the login page.
3. If the HTTP request doesn't get a response at all, `renewAWGToken` throws, which probably causes something bad to happen, but the situation is likely unrecoverable at that point anyway.

If so, I think 1 and 2 are fine, but to Phil's point, it'd probably be better to cache false in 3 so future calls return the cached failure. Depending on how the exception is handled, though, it might not matter.

Reply:
Hm! I hadn't understood 3! You're both 100% right.
For documentation's sake, if the request fails, I'm passing the error response from the server (which could be anything, e.g. an object, a string or an Error entity, all of which are truthy).
Setting the queryCache to any of those will be true, and result in allowing the subsequent API call which this token check should disallow. That other request in turn will reject, and thus handling the logout.

Though that manages the result as per spec (user is logged out of the UI forcefully), this check is not preventing the other API call.